io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,356 @@
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*
+*/
+
+#include <pcl/io/eigen.h>
+#include <pcl/io/boost.h>
+#include <pcl/io/grabber.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+
+#include <librealsense2/rs.hpp>
+#include <pcl/io/real_sense_2_grabber.h>
+#include <pcl/common/time.h>
+
+namespace pcl
+{
+
+  RealSense2Grabber::RealSense2Grabber (const uint32_t serial_number)
+    : running_ (false)
+    , quit_ (false)
+    , serial_number_ (serial_number)
+    , signal_PointXYZ (nullptr)
+    , signal_PointXYZI (nullptr)
+    , signal_PointXYZRGB (nullptr)
+    , signal_PointXYZRGBA (nullptr)
+    , fps_ (0)
+    , device_width_ (424)
+    , device_height_ (240)
+    , target_fps_ (30)
+  {
+    signal_PointXYZ = createSignal<signal_librealsense_PointXYZ> ();
+    signal_PointXYZI = createSignal<signal_librealsense_PointXYZI> ();
+    signal_PointXYZRGB = createSignal<signal_librealsense_PointXYZRGB> ();
+    signal_PointXYZRGBA = createSignal<signal_librealsense_PointXYZRGBA> ();
+  }
+
+  RealSense2Grabber::RealSense2Grabber (const std::string& file_name)
+    : file_name_(file_name)
+  {
+    RealSense2Grabber ();
+  }
+
+
+  RealSense2Grabber::~RealSense2Grabber ()
+  {
+    stop ();
+
+    disconnect_all_slots<signal_librealsense_PointXYZ> ();
+    disconnect_all_slots<signal_librealsense_PointXYZI> ();
+    disconnect_all_slots<signal_librealsense_PointXYZRGB> ();
+    disconnect_all_slots<signal_librealsense_PointXYZRGBA> ();
+
+    thread_.join ();
+
+    pipe_.stop ();
+  }
+
+  void 
+  RealSense2Grabber::start ()
+  {
+    if (isRunning ())
+      return;
+
+    running_ = true;
+
+    rs2::config cfg;
+
+    // capture from file
+    if (!file_name_.empty ())
+    {
+      cfg.enable_device_from_file (file_name_);
+    }
+    else
+    {
+      if (serial_number_)
+        cfg.enable_device (std::to_string (serial_number_));
+
+      cfg.enable_stream (RS2_STREAM_COLOR, device_width_, device_height_, RS2_FORMAT_RGB8, target_fps_);
+      cfg.enable_stream (RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_ANY, target_fps_);
+      cfg.enable_stream (RS2_STREAM_INFRARED, device_width_, device_height_, RS2_FORMAT_ANY, target_fps_);
```

Comment:
I think that it is safety to specify format type explicitly.
I think you are assuming <code>UYVY</code> in the Infrared frame. If <code>Y8</code> is specified, conversion (from UYVY to Y8) becomes unnecessary in <code>convertInfraredDepthToPointXYZI()</code>.

```patch
- cfg.enable_stream (RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_ANY, target_fps_);
+ cfg.enable_stream (RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_Z16, target_fps_);
- cfg.enable_stream (RS2_STREAM_INFRARED, device_width_, device_height_, RS2_FORMAT_ANY, target_fps_);
+ cfg.enable_stream (RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_Y8, target_fps_);
```

Reply:
That makes sense.  I have to see how the intensity data will look like for that case